cpp/src/copying/pack.cpp

```diff
@@ -156,27 +156,46 @@ std::vector<uint8_t> pack_metadata(ColumnIter begin,
                                    uint8_t const* contiguous_buffer,
                                    size_t buffer_size)
 {
-  std::vector<serialized_column> metadata;
+  auto mb = metadata_builder(std::distance(begin, end));
 
-  // first metadata entry is a stub indicating how many total (top level) columns
-  // there are
-  metadata.emplace_back(
-    data_type{type_id::EMPTY}, static_cast<size_type>(std::distance(begin, end)), 0, -1, -1, 0);
-
-  std::for_each(begin, end, [&metadata, &contiguous_buffer, &buffer_size](column_view const& col) {
-    build_column_metadata(metadata, col, contiguous_buffer, buffer_size);
+  std::for_each(begin, end, [&mb, &contiguous_buffer, &buffer_size](column_view const& col) {
+    build_column_metadata(mb, col, contiguous_buffer, buffer_size);
   });
 
-  // convert to anonymous bytes
-  std::vector<uint8_t> metadata_bytes;
-  auto const metadata_begin = reinterpret_cast<uint8_t const*>(metadata.data());
-  std::copy(metadata_begin,
-            metadata_begin + (metadata.size() * sizeof(serialized_column)),
-            std::back_inserter(metadata_bytes));
-
-  return metadata_bytes;
+  return mb.build();
 }
 
+class metadata_builder_impl {
+ public:
+  metadata_builder_impl() = default;
+
+  void add_column_to_meta(data_type col_type,
+                          size_type col_size,
+                          size_type col_null_count,
+                          int64_t data_offset,
+                          int64_t null_mask_offset,
+                          size_type num_children)
+  {
+    metadata.emplace_back(
+      col_type, col_size, col_null_count, data_offset, null_mask_offset, num_children);
+  }
+
+  std::vector<uint8_t> build()
+  {
+    // convert to anonymous bytes
+    std::vector<uint8_t> metadata_bytes;
```

Comment:
So `std::byte` isn't defined in cython. @ttnghia had a similar suggestion in my prior PR https://github.com/rapidsai/cudf/pull/13186 and we ended up settling for `uint8_t` because otherwise the python code couldn't use this field.

Reply:
Oh, wow. I hadn't considered that. TIL.